src/workerd/jsg/value.h

```diff
@@ -468,18 +469,26 @@ class StringWrapper {
       kj::Maybe<v8::Local<v8::Object>> parentObject) {
     // TODO(cleanup): Move to a HeaderStringWrapper in the api directory.
     v8::Local<v8::String> str = check(handle->ToString(context));
-    auto result =
-        ByteString(KJ_ASSERT_NONNULL(tryUnwrap(context, str, (kj::String*)nullptr, parentObject)));
-    if (str->ContainsOnlyOneByte()) {
-      auto buf = kj::heapArray<kj::byte>(str->Length() + 1);
-      str->WriteOneByte(context->GetIsolate(), buf.begin());
-      for (auto b: buf) {
-        if (b >= 128) {
-          result.warning = ByteString::Warning::CONTAINS_EXTENDED_ASCII;
-          break;
-        }
+    auto result = ByteString(KJ_ASSERT_NONNULL(
+        tryUnwrap(context, str, static_cast<kj::String*>(nullptr), parentObject)));
+
+    v8::String::ValueView str_view(context->GetIsolate(), str);
+
+    if (str_view.is_one_byte()) [[likely]] {
+      // Storage is one-byte.
+      if (!simdutf::validate_ascii(
+              reinterpret_cast<const char*>(str_view.data8()), str_view.length())) [[unlikely]] {
+        result.warning = ByteString::Warning::CONTAINS_EXTENDED_ASCII;
+      }
+    } else if (str->ContainsOnlyOneByte()) {
+      // Storage is two-byte, but it contains only one byte.
+      KJ_ASSERT(!str_view.is_one_byte());
+      auto buf = reinterpret_cast<const char*>(str_view.data16());
```

Comment:
Err... my first impression was that this cast obviously doesn't work, you can't just reinterpret a pointer from a char16 array to a char8 array.

But I guess it _does_ work in that we already know that all the actual characters are 8-bit, so basically this is an array where every other byte is NUL, which is ASCII, and it's only the non-NUL chars that you intend to validate?

But then you need to double `str_view.length()` on the next line, right? And all this deserves a massive comment.

How about instead, just use the result buffer?

`validate_ascii(result.begin(), result.size())`

The result is UTF-8 encoded, but validate_ascii should handle that just fine, presumably?

In fact, you could do the above `validate_ascii` check first, and only call `ContainsOnlyOneByte()` in the case that it fails (to decide which warning to return), to optimize for the success case...

Reply:
Actually, yes. We're over-engineering it. Directly calling simdutf and eliminating the happy path in the beginning makes things a lot simpler to understand and comprehend in the first look.